watchtower-plugin/src/dbm.rs

```diff
@@ -139,24 +140,25 @@ impl DBM {
         .map_err(|_| Error::NotFound)
     }
 
-    /// Stores a tower record into the database.
+    /// Stores a tower record into the database alongside the corresponding registration receipt.
     pub fn store_tower_record(
-        &self,
+        &mut self,
         tower_id: TowerId,
         net_addr: &str,
         receipt: &RegistrationReceipt,
     ) -> Result<(), Error> {
-        let query =
-            "INSERT OR REPLACE INTO towers (tower_id, net_addr, available_slots, subscription_expiry) VALUES (?1, ?2, ?3, ?4)";
-        self.store_data(
-            query,
-            params![
-                tower_id.to_vec(),
-                net_addr,
-                receipt.available_slots(),
-                receipt.subscription_expiry()
-            ],
+        let tx = self.get_mut_connection().transaction().unwrap();
+        tx.execute(
+            "INSERT INTO towers (tower_id, net_addr, available_slots) VALUES (?1, ?2, ?3) ON CONFLICT (tower_id) DO UPDATE SET net_addr = ?2, available_slots = ?3",
+            params![tower_id.to_vec(), net_addr, receipt.available_slots()],
         )
+        .map_err( Error::Unknown)?;
+        // FIXME: This may need a guard. What if a tower returns the same end_time twice?
```

Comment:
Yeah we better send a combined receipt from the tower.
If we know that our sub isn't over yet, we should verify the validity of the received receipt and if it's fine we replace the old one by it.
This case we won't actually need to keep a `registration_id`.

I think we can also have `tower_id` only be the PK, because, why would we need to store old receipts?

Reply:
This holds as long as there is no gap in the subscriptions:
```
|------------------|---------------------|
	S1		    S2
```

If the client has an appointment receipt for `A1` sent to the tower during `S1`, and there is no gap between `S1` and `S2`, then the user cannot trick the tower, having `S2` would suffice.

```
|------------------|  GAP  |---------------------|
	S1			     S2
```

If there is a gap between `S1` and `S2`, if we allow `S2` and `A1` to be enough to prove misbehaving, then the user can wait until its data is wiped from the tower (GAP), register again (`S2`) and trigger `A1`. The tower won't be able to react to `A1`.

Therefore, the misbehaving proof should include registration receipts from `A1` start_block to the block the breach was not penalized (without gaps). If the user does not interrupt the subscription, then the tower can simply batch the receipts and both will only need to store a single one.

LMK if this logic holds or if you find anything weird.